Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,995 @@
+`F(S) = F(S ^ (q ^ n))` for any natural number `n`. -/
+theorem adjoin_eq_adjoin_pow_expChar_pow_of_isSeparable (S : Set E) [IsSeparable F (adjoin F S)]
+    (q : ℕ) [ExpChar F q] (n : ℕ) : adjoin F S = adjoin F ((fun x ↦ x ^ q ^ n) '' S) := by
+  set L := adjoin F S
+  set M := adjoin F ((fun x ↦ x ^ q ^ n) '' S)
+  have hi : M ≤ L := by
+    rw [adjoin_le_iff]
+    rintro _ ⟨y, hy, rfl⟩
+    exact pow_mem (subset_adjoin F S hy) _
+  letI := (inclusion hi).toAlgebra
+  letI : Module M L := Algebra.toModule
+  letI : SMul M L := Algebra.toSMul
+  haveI : IsScalarTower F M L := IsScalarTower.of_algebraMap_eq (congrFun rfl)
+  haveI : IsSeparable M (extendScalars hi) := isSeparable_tower_top_of_isSeparable F M L
+  haveI : IsPurelyInseparable M (extendScalars hi) := by
+    haveI := expChar_of_injective_algebraMap (algebraMap F M).injective q
+    rw [extendScalars_adjoin hi, isPurelyInseparable_adjoin_iff_pow_mem M _ q]
+    exact fun x hx ↦ ⟨n, ⟨x ^ q ^ n, subset_adjoin F _ ⟨x, hx, rfl⟩⟩, rfl⟩
+  simpa only [extendScalars_restrictScalars, restrictScalars_bot_eq_self] using congr_arg
+    (restrictScalars F) (extendScalars hi).eq_bot_of_isPurelyInseparable_of_isSeparable
+
+/-- If `E / F` is a separable field extension of exponential characteristic `q`, then
+`F(S) = F(S ^ (q ^ n))` for any subset `S` of `E` and any natural number `n`. -/
+theorem adjoin_eq_adjoin_pow_expChar_pow_of_isSeparable' [IsSeparable F E] (S : Set E)
+    (q : ℕ) [ExpChar F q] (n : ℕ) : adjoin F S = adjoin F ((fun x ↦ x ^ q ^ n) '' S) :=
+  haveI := isSeparable_tower_bot_of_isSeparable F (adjoin F S) E
+  adjoin_eq_adjoin_pow_expChar_pow_of_isSeparable F E S q n
+
+-- TODO: prove the converse when `F(S) / F` is finite
+/-- If `F` is a field of exponential characteristic `q`, `F(S) / F` is separable, then
+`F(S) = F(S ^ q)`. -/
+theorem adjoin_eq_adjoin_pow_expChar_of_isSeparable (S : Set E) [IsSeparable F (adjoin F S)]
+    (q : ℕ) [ExpChar F q] : adjoin F S = adjoin F ((fun x ↦ x ^ q) '' S) :=
+  pow_one q ▸ adjoin_eq_adjoin_pow_expChar_pow_of_isSeparable F E S q 1
+
+/-- If `E / F` is a separable field extension of exponential characteristic `q`, then
+`F(S) = F(S ^ q)` for any subset `S` of `E`. -/
+theorem adjoin_eq_adjoin_pow_expChar_of_isSeparable' [IsSeparable F E] (S : Set E)
+    (q : ℕ) [ExpChar F q] : adjoin F S = adjoin F ((fun x ↦ x ^ q) '' S) :=
+  pow_one q ▸ adjoin_eq_adjoin_pow_expChar_pow_of_isSeparable' F E S q 1
+
+end IntermediateField
+
+variable {F E} in
+/-- If `E / F` is a separable extension of exponential characteristic `q`, if `{ u_i }` is a family
+of elements of `E` which `F`-linearly spans `E`, then `{ u_i ^ (q ^ n) }` also `F`-linearly spans
+`E` for any natural number `n`. -/
+theorem Field.span_map_pow_expChar_pow_eq_top_of_isSeparable (q : ℕ) [hF : ExpChar F q] (n : ℕ)
+    [IsSeparable F E] {ι : Type*} {v : ι → E} (h : Submodule.span F (Set.range v) = ⊤) :
+    Submodule.span F (Set.range ((fun x ↦ x ^ q ^ n) ∘ v)) = ⊤ := by
```

Comment:
I've opened an easy PR #10017 and a refactor PR #10016 with the main purpose of golfing this proof.

Reply:
The golfe proof is [here](https://github.com/leanprover-community/mathlib4/commit/3f93a2df3021f115939d0c9394c3efa6d0c31183).

Reply:
I have [more golfs](https://github.com/leanprover-community/mathlib4/compare/e5eb62dff7bb69f587ba86add3fe9f59a7405cf7...purely_insep_merge) at that branch but other than that I think this is ready to go!

Reply:
Would you also merge my branch?

Reply:
> Would you also merge my branch?

Sure, thank you very much. Would you like to push your changes to this PR by yourself?

Reply:
Done! Hopefully CI is happy ...